### ISSUE TITLE
Avoid "except Exception" or re-raise them when possible

### ISSUE THREAD

**Issue by achimnol** [aio-libs/sockjs#35]
I have run into some nasty corner cases where my code is "slightly" broken but cannot see what's wrong.
Currently many places (example [1](https://github.com/aio-libs/sockjs/blob/b5919ac/sockjs/session.py#L101) and [2](https://github.com/aio-libs/sockjs/blob/34f4c4d/sockjs/route.py#L152)) catches "Exception" and wraps the exception information into CLOSE frames with exception logging.
If this is combined with [a missing argument in `onclose()`](https://github.com/aio-libs/sockjs/blob/546438a/examples/chat.html#L36), which may happen quite frequently due to the lack of documentation/example in [the client-side Javascript](https://github.com/sockjs/sockjs-client), it is hard to debug what is the real problem. (In fact, client-side `onclose()` handler gets a pair of error code and message. But nowhere this is described in the client docs!)
In my case, it was even harder because my application code was running under a custom logger setting which only exposed specific logging namespaces but "sockjs".

I think it is always a better practice to leave handling "any kind of exception" for application writers instead of library writers. Libraries should _catch only specific exceptions and re-raise others_ after doing their own clean-ups, whenever possible. Otherwise, such behavior should be explicitly documented.

**Comment by asvetlov** [comment]
Web-handlers are not called by user code but in reaction to incoming connection acceptance.
Thus there is no chance to bubble-up exception to proper handler.
The best that a handler can do is logging exception, changing current state and sending proper HTTP status code back.
That's exactly what SockJS does.

The problem is not specific to SockJS, any web-server has the same reactive behavior.

P.S.
If you have a concrete improving please don't hesitate to make a Pull Request.

**Comment by achimnol** [comment]
I understand that passively executed web handlers have no place to bubble up exceptions.
However, in my case the root cause was my fault: passing bytes instead of strings to `session.send()` in the server-side code. This has caused an exception in [the example 1](https://github.com/aio-libs/sockjs/blob/b5919ac/sockjs/session.py#L101).

**Comment by asvetlov** [comment]
If you blame that a method doesn't check parameters types -- please make a PR for fixing it.

**Comment by achimnol** [comment]
After working with this for a while, I concluded that it is not simple to pass exceptions raised in library codes that does not reside in the direct stack called by the library user (i.e., in asynchronous tasks) to the user code.
Maybe we need explicit user-defined callbacks for such "asynchronous" exceptions.
Currently I'm not very sure how we could implement this on sockjs with a clean interface. Need some time to think.

**Comment by asvetlov** [comment]
asyncio uses `loop.call_exception_handler()`, looks like we need it too.

**Comment by achimnol** [comment]
Ah, I finally realized that my issue is actually related to python/asyncio#397!
After writing several test cases including trying out `loop.set_exception_handler`, I found that exception swallowing occurs when the task I spawn to receive messages from another server and then relay them via sockjs raises exceptions, not when the incoming message handler does. This means that current exception handling in sockjs is not the root cause of my problems.

I'm closing this issue. :)
